File: book/petrol-pump/c#/PetrolPump/Chapter4/Section7/Fill.cs
using PetrolPump.Chapter4.Section6;
using Sodium.Frp;
using Sodium.Functional;

namespace PetrolPump.Chapter4.Section7
{
    internal class Fill
    {
        public Fill(
            Stream<Unit> sClearAccumulator, Stream<int> sFuelPulses,
            Cell<double> calibration, Cell<double> price1,
            Cell<double> price2, Cell<double> price3,
            Stream<Fuel> sStart)
        {
            this.Price = CapturePrice(sStart, price1, price2, price3);
            this.LitersDelivered = AccumulatePulsesPump.Accumulate(
                sClearAccumulator, sFuelPulses, calibration);
            this.DollarsDelivered = this.LitersDelivered.Lift(this.Price,
                (liters, price) => liters * price);
        }

        public static Cell<double> CapturePrice(
            Stream<Fuel> sStart,
            Cell<double> price1, Cell<double> price2,
            Cell<double> price3)
        {
            Stream<double> sPrice1 = sStart.Snapshot(price1,
                (f, p) => f == Fuel.One ? Maybe.Some(p) : Maybe.None).FilterMaybe();
            Stream<double> sPrice2 = sStart.Snapshot(price2,
                (f, p) => f == Fuel.Two ? Maybe.Some(p) : Maybe.None).FilterMaybe();
            Stream<double> sPrice3 = sStart.Snapshot(price3,
                (f, p) => f == Fuel.Three ? Maybe.Some(p) : Maybe.None).FilterMaybe();

            return sPrice1.OrElse(sPrice2.OrElse(sPrice3)).Hold(0.0);
        }

        public Cell<double> Price { get; }
        public Cell<double> DollarsDelivered { get; }
        public Cell<double> LitersDelivered { get; }
    }
}

File: book/petrol-pump/java/chapter4/section7/Fill.java
package chapter4.section7;

import pump.*;
import chapter4.section6.AccumulatePulsesPump;
import nz.sodium.*;
import java.util.Optional;

public class Fill {
    public final Cell<Double> price;
    public final Cell<Double> dollarsDelivered;
    public final Cell<Double> litersDelivered;

    public Fill(
            Stream<Unit> sClearAccumulator, Stream<Integer> sFuelPulses,
            Cell<Double> calibration, Cell<Double> price1,
            Cell<Double> price2, Cell<Double> price3,
            Stream<Fuel> sStart) {
        price = capturePrice(sStart, price1, price2, price3);
        litersDelivered = AccumulatePulsesPump.accumulate(
                sClearAccumulator, sFuelPulses, calibration);
        dollarsDelivered = litersDelivered.lift(price,
                (liters, price_) -> liters * price_);
    }

    public static Cell<Double> capturePrice(
            Stream<Fuel> sStart,
            Cell<Double> price1, Cell<Double> price2,
            Cell<Double> price3) {
        Stream<Double> sPrice1 = Stream.filterOptional(
            sStart.snapshot(price1,
                (f, p) -> f == Fuel.ONE ? Optional.of(p)
                                        : Optional.empty()));
        Stream<Double> sPrice2 = Stream.filterOptional(
            sStart.snapshot(price2,
                (f, p) -> f == Fuel.TWO ? Optional.of(p)
                                        : Optional.empty()));
        Stream<Double> sPrice3 = Stream.filterOptional(
            sStart.snapshot(price3,
                (f, p) -> f == Fuel.THREE ? Optional.of(p)
                                          : Optional.empty()));

        return sPrice1.orElse(sPrice2.orElse(sPrice3))
                      .hold(0.0);
    }
}
